Add ex_fetch_power_state option to list_nodes().

tetron opened the pull request:
## Add ex_fetch_power_state option to list_nodes().

### Description

Allows caller to skip (potentially expensive) requests for extended node
power state status.

Also set timeout on connection object used to fetch token.

### Status

Ready for review

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation
- [x] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

Commit:
Add ex_fetch_status option to list_nodes().

Allows caller to skip (potentially expensive) requests for extended node
status.

Also set timeout on connection object used to fetch token.

tfmorris commented on libcloud/compute/drivers/azure_arm.py:
```diff
@@ -371,6 +371,9 @@ def list_nodes(self, ex_resource_group=None, ex_fetch_nic=True):
         IP address information for nodes (requires extra API calls).
         :type ex_urn: ``bool``
 
+        :param ex_fetch_status: Fetch node instance status (requires extra API calls).
```
Suggest this be phrased as "requires an extra API call per node"

tfmorris commented on libcloud/compute/drivers/azure_arm.py:
```diff
@@ -371,6 +371,9 @@ def list_nodes(self, ex_resource_group=None, ex_fetch_nic=True):
         IP address information for nodes (requires extra API calls).
```
Suggest this be phrased as "requires an extra API call for each NIC"

tfmorris commented on libcloud/compute/drivers/azure_arm.py:
```diff
@@ -1942,39 +1945,40 @@ def _to_node(self, data, fetch_nic=True):
+                        state = NodeState.UPDATING
+                        break
+                    elif status["code"] == "ProvisioningState/succeeded":
+                        pass
```
The ProvisioningState is available from the virtual machines list resource group response: https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/virtualmachines-list-resource-group
without having to make an extra request.

I think `"provisioningState": "succeeded"` is close enough to return NodeState.RUNNING even without the additional detail of PowerState.

pquentin commented:
@tfmorris Thank you for the review!

@tetron Can you please fix the lint, possibly by moving the status fetch in another function?

Commit:
Rename fetch_status to fetch_power_state.

Return node state based on 'provisioningState' if fetch_power_state is False.
Update API documentation.

Commit:
Fix provisioningState.  Fix lint.

Commit:
Force provisioningState to lowercase for comparison.

tetron commented:
It's been 20 minute since I pushed and Travis hasn't started running tests yet.  I don't know if it is a rate limit thing or Travis is having trouble (but currently no downtime is reported on their status page).  Guess I'll try to be patient.

pquentin commented:
@tetron Heh. It did take some time to run, and there's one lint issue remaining. You can use `tox -e lint` to get results faster!

Commit:
Fix lint.

pquentin commented:
@tetron Thanks! We just need tests for the `ex_fetch_power_state=False` branch to ensure coverage does not decrease.

ldipenti committed:
Added test for ex_fetch_power_state=False

pquentin commented:
Thanks @tetron @ldipenti this is now merged in trunk.

ldipenti commented:
Great, thanks @pquentin !
